File: src/dotnet/CarbonOps.Parser.Contracts/ProductionConfigBoundary.cs
using System.Text.Json;

namespace CarbonOps.Parser.Contracts;

public sealed record ProductionConfigValidationIssue(
    string Code,
    string Message,
    string FieldName,
    string Severity = "error");

public sealed record ProductionConfigValidationResult
{
    public IReadOnlyList<ProductionConfigValidationIssue> Issues { get; }

    public bool IsValid => Issues.Count == 0;

    public ProductionConfigValidationResult(
        IEnumerable<ProductionConfigValidationIssue>? issues = null)
    {
        Issues = Array.AsReadOnly((issues ?? []).ToArray());
    }
}

public sealed record ProductionConfigBoundaryDescription(
    IReadOnlyList<string> RequiredEnvironmentVariables,
    IReadOnlyList<string> SecretEnvironmentVariables,
    string Provider,
    bool LoadsEnvironment,
    bool LoadsConfigFiles,
    bool LoadsCredentials,
    bool LogsSecretValues,
    IReadOnlyList<string> Notes);

public static class ProductionConfigBoundary
{
    public const string RawPostgreSQLConnectionStringKey = "CARBONOPS_PARSER_POSTGRES_CONNECTION_STRING";

    public static readonly IReadOnlyList<string> RequiredEnvironmentVariables = Array.AsReadOnly(new[]
    {
        "CARBONOPS_PARSER_ENV",
        "CARBONOPS_PARSER_DATABASE_PROVIDER",
        "CARBONOPS_PARSER_POSTGRES_HOST",
        "CARBONOPS_PARSER_POSTGRES_PORT",
        "CARBONOPS_PARSER_POSTGRES_DATABASE",
        "CARBONOPS_PARSER_POSTGRES_USERNAME",
        "CARBONOPS_PARSER_POSTGRES_PASSWORD",
        "CARBONOPS_PARSER_POSTGRES_SCHEMA",
        "CARBONOPS_PARSER_RAW_ARCHIVE_PATH",
        "CARBONOPS_PARSER_LOG_LEVEL",
    });

    public static readonly IReadOnlyList<string> SecretEnvironmentVariables = Array.AsReadOnly(new[]
    {
        "CARBONOPS_PARSER_POSTGRES_PASSWORD",
    });

    public static readonly IReadOnlyList<string> KnownConfigurationKeys = Array.AsReadOnly(
        RequiredEnvironmentVariables
            .Append(RawPostgreSQLConnectionStringKey)
            .ToArray());

    private static readonly HashSet<string> ValidLogLevels = new(StringComparer.OrdinalIgnoreCase)
    {
        "debug",
        "info",
        "warning",
        "error",
        "critical",
    };

    public static ProductionConfigBoundaryDescription Describe() =>
        new(
            RequiredEnvironmentVariables,
            SecretEnvironmentVariables,
            "postgres",
            LoadsEnvironment: true,
            LoadsConfigFiles: true,
            LoadsCredentials: true,
            LogsSecretValues: false,
            [
                "The .NET service entrypoint loads an optional explicit JSON config file and process environment.",
                "Environment values override config file values for known production keys.",
                "CARBONOPS_PARSER_POSTGRES_PASSWORD is required but never returned.",
                "Connection strings are not accepted as production config values.",
                "Validation messages name keys only and do not echo configured values.",
            ]);

    public static ProductionConfigValidationResult Validate(
        IReadOnlyDictionary<string, string?> values)
    {
        var issues = new List<ProductionConfigValidationIssue>();

        foreach (var envVar in RequiredEnvironmentVariables)
        {
            if (!HasText(Get(values, envVar)))
            {
                issues.Add(new ProductionConfigValidationIssue(
                    "PRODUCTION_CONFIG_MISSING_REQUIRED_ENV_VAR",
                    $"{envVar} must be set for production startup.",
                    envVar));
            }
        }

        var provider = Get(values, "CARBONOPS_PARSER_DATABASE_PROVIDER");
        if (provider is not null &&
            HasText(provider) &&
            !string.Equals(provider.Trim(), "postgres", StringComparison.OrdinalIgnoreCase))
        {
            issues.Add(new ProductionConfigValidationIssue(
                "PRODUCTION_CONFIG_UNSUPPORTED_DATABASE_PROVIDER",
                "Unsupported database provider. Phase 1 supports postgres only.",
                "CARBONOPS_PARSER_DATABASE_PROVIDER"));
        }

        ValidatePort(Get(values, "CARBONOPS_PARSER_POSTGRES_PORT"), issues);
        ValidateLogLevel(Get(values, "CARBONOPS_PARSER_LOG_LEVEL"), issues);

        if (HasText(Get(values, RawPostgreSQLConnectionStringKey)))
        {
            issues.Add(new ProductionConfigValidationIssue(
                "PRODUCTION_CONFIG_RAW_CONNECTION_STRING_NOT_ALLOWED",
                "Raw PostgreSQL connection strings are not accepted; use split non-secret fields and CARBONOPS_PARSER_POSTGRES_PASSWORD.",
                RawPostgreSQLConnectionStringKey));
        }

        return new ProductionConfigValidationResult(issues);
    }

    private static void ValidatePort(
        string? rawValue,
        ICollection<ProductionConfigValidationIssue> issues)
    {
        if (!HasText(rawValue))
        {
            return;
        }

        if (!int.TryParse(rawValue, out var port) || port is < 1 or > 65535)
        {
            issues.Add(new ProductionConfigValidationIssue(
                "PRODUCTION_CONFIG_INVALID_POSTGRES_PORT",
                "CARBONOPS_PARSER_POSTGRES_PORT must be an integer between 1 and 65535.",
                "CARBONOPS_PARSER_POSTGRES_PORT"));
        }
    }

    private static void ValidateLogLevel(
        string? rawValue,
        ICollection<ProductionConfigValidationIssue> issues)
    {
        if (!HasText(rawValue))
        {
            return;
        }

        if (rawValue is not null && !ValidLogLevels.Contains(rawValue.Trim()))
        {
            issues.Add(new ProductionConfigValidationIssue(
                "PRODUCTION_CONFIG_INVALID_LOG_LEVEL",
                "CARBONOPS_PARSER_LOG_LEVEL must be one of debug, info, warning, error, or critical.",
                "CARBONOPS_PARSER_LOG_LEVEL"));
        }
    }

    private static bool HasText(string? value) => !string.IsNullOrWhiteSpace(value);

    private static string? Get(IReadOnlyDictionary<string, string?> values, string key) =>
        values.TryGetValue(key, out var value) ? value : null;
}

public sealed record ProductionConfigLoadResult(
    IReadOnlyDictionary<string, string?> Values,
    IReadOnlyList<ProductionConfigValidationIssue> Issues,
    bool ConfigFileLoaded,
    bool EnvironmentLoaded);

public static class ProductionConfigLoader
{
    public static ProductionConfigLoadResult Load(
        string? configPath,
        IReadOnlyDictionary<string, string?> environment)
    {
        var values = new Dictionary<string, string?>(StringComparer.Ordinal);
        var issues = new List<ProductionConfigValidationIssue>();
        var configFileLoaded = false;

        foreach (var key in ProductionConfigBoundary.KnownConfigurationKeys)
        {
            values[key] = null;
        }

        if (!string.IsNullOrWhiteSpace(configPath))
        {
            configFileLoaded = LoadFileValues(configPath, values, issues);
        }

        foreach (var key in ProductionConfigBoundary.KnownConfigurationKeys)
        {
            if (environment.TryGetValue(key, out var value))
            {
                values[key] = value;
            }
        }

        return new ProductionConfigLoadResult(
            values,
            Array.AsReadOnly(issues.ToArray()),
            configFileLoaded,
            EnvironmentLoaded: true);
    }

    private static bool LoadFileValues(
        string configPath,
        IDictionary<string, string?> values,
        ICollection<ProductionConfigValidationIssue> issues)
    {
        if (!File.Exists(configPath))
        {
            issues.Add(new ProductionConfigValidationIssue(
                "PRODUCTION_CONFIG_FILE_NOT_FOUND",
                "Config file was not found.",
                "config"));
            return false;
        }

        try
        {
            using var stream = File.OpenRead(configPath);
            using var document = JsonDocument.Parse(stream);

            if (document.RootElement.ValueKind != JsonValueKind.Object)
            {
                issues.Add(new ProductionConfigValidationIssue(
                    "PRODUCTION_CONFIG_FILE_INVALID_SHAPE",
                    "Config file root must be a JSON object.",
                    "config"));
                return false;
            }

            foreach (var property in document.RootElement.EnumerateObject().OrderBy(item => item.Name, StringComparer.Ordinal))
            {
                if (!values.ContainsKey(property.Name))
                {
                    continue;
                }

                if (!TryReadScalar(property.Value, out var value))
                {
                    issues.Add(new ProductionConfigValidationIssue(
                        "PRODUCTION_CONFIG_FILE_INVALID_VALUE",
                        "Config file values must be scalar strings, numbers, booleans, or null.",
                        property.Name));
                    continue;
                }

                values[property.Name] = value;
            }

            return true;
        }
        catch (JsonException)
        {
            issues.Add(new ProductionConfigValidationIssue(
                "PRODUCTION_CONFIG_FILE_INVALID_JSON",
                "Config file must be valid JSON.",
                "config"));
            return false;
        }
        catch (IOException)
        {
            issues.Add(new ProductionConfigValidationIssue(
                "PRODUCTION_CONFIG_FILE_READ_FAILED",
                "Config file could not be read.",
                "config"));
            return false;
        }
        catch (UnauthorizedAccessException)
        {
            issues.Add(new ProductionConfigValidationIssue(
                "PRODUCTION_CONFIG_FILE_READ_FAILED",
                "Config file could not be read.",
                "config"));
            return false;
        }
    }

    private static bool TryReadScalar(JsonElement element, out string? value)
    {
        switch (element.ValueKind)
        {
            case JsonValueKind.String:
                value = element.GetString();
                return true;
            case JsonValueKind.Number:
            case JsonValueKind.True:
            case JsonValueKind.False:
                value = element.GetRawText();
                return true;
            case JsonValueKind.Null:
                value = null;
                return true;
            default:
                value = null;
                return false;
        }
    }
}

File: src/carbonfactor_parser/persistence/production_config_boundary.py
"""Production runtime configuration boundary without environment loading."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Mapping


PRODUCTION_CONFIG_REQUIRED_ENV_VARS: tuple[str, ...] = (
    "CARBONOPS_PARSER_ENV",
    "CARBONOPS_PARSER_DATABASE_PROVIDER",
    "CARBONOPS_PARSER_POSTGRES_HOST",
    "CARBONOPS_PARSER_POSTGRES_PORT",
    "CARBONOPS_PARSER_POSTGRES_DATABASE",
    "CARBONOPS_PARSER_POSTGRES_USERNAME",
    "CARBONOPS_PARSER_POSTGRES_PASSWORD",
    "CARBONOPS_PARSER_POSTGRES_SCHEMA",
    "CARBONOPS_PARSER_RAW_ARCHIVE_PATH",
    "CARBONOPS_PARSER_LOG_LEVEL",
)

PRODUCTION_CONFIG_SECRET_ENV_VARS: tuple[str, ...] = (
    "CARBONOPS_PARSER_POSTGRES_PASSWORD",
)

_VALID_LOG_LEVELS = frozenset({"debug", "info", "warning", "error", "critical"})


@dataclass(frozen=True)
class ProductionConfigValidationIssue:
    """Safe production config validation issue with no configured values."""

    code: str
    message: str
    field_name: str
    severity: str = "error"


@dataclass(frozen=True)
class ProductionConfigValidationResult:
    """Side-effect-free validation result for production config mappings."""

    issues: tuple[ProductionConfigValidationIssue, ...] = ()

    @property
    def is_valid(self) -> bool:
        return not self.issues


@dataclass(frozen=True)
class ProductionConfigBoundaryDescription:
    """Shared Python/.NET production configuration expectations."""

    required_env_vars: tuple[str, ...]
    secret_env_vars: tuple[str, ...]
    provider: str
    loads_environment: bool
    loads_config_files: bool
    loads_credentials: bool
    logs_secret_values: bool
    notes: tuple[str, ...]


def describe_production_config_boundary() -> ProductionConfigBoundaryDescription:
    """Describe production config expectations without reading configuration."""

    return ProductionConfigBoundaryDescription(
        required_env_vars=PRODUCTION_CONFIG_REQUIRED_ENV_VARS,
        secret_env_vars=PRODUCTION_CONFIG_SECRET_ENV_VARS,
        provider="postgres",
        loads_environment=False,
        loads_config_files=False,
        loads_credentials=False,
        logs_secret_values=False,
        notes=(
            "Callers pass an explicit mapping for validation.",
            "CARBONOPS_PARSER_POSTGRES_PASSWORD is required but never returned.",
            "Connection strings are not accepted as production config values.",
            "Validation messages name keys only and do not echo configured values.",
        ),
    )


def validate_production_config_mapping(
    values: Mapping[str, str | None],
) -> ProductionConfigValidationResult:
    """Validate required production config keys without loading env or secrets."""

    issues: list[ProductionConfigValidationIssue] = []

    for env_var in PRODUCTION_CONFIG_REQUIRED_ENV_VARS:
        if not _has_text(values.get(env_var)):
            issues.append(
                ProductionConfigValidationIssue(
                    code="PRODUCTION_CONFIG_MISSING_REQUIRED_ENV_VAR",
                    message=f"{env_var} must be set for production startup.",
                    field_name=env_var,
                ),
            )

    provider = values.get("CARBONOPS_PARSER_DATABASE_PROVIDER")
    if _has_text(provider) and provider.strip().lower() != "postgres":
        issues.append(
            ProductionConfigValidationIssue(
                code="PRODUCTION_CONFIG_UNSUPPORTED_DATABASE_PROVIDER",
                message="Unsupported database provider. Phase 1 supports postgres only.",
                field_name="CARBONOPS_PARSER_DATABASE_PROVIDER",
            ),
        )

    _validate_port(values.get("CARBONOPS_PARSER_POSTGRES_PORT"), issues)
    _validate_log_level(values.get("CARBONOPS_PARSER_LOG_LEVEL"), issues)

    if _has_text(values.get("CARBONOPS_PARSER_POSTGRES_CONNECTION_STRING")):
        issues.append(
            ProductionConfigValidationIssue(
                code="PRODUCTION_CONFIG_RAW_CONNECTION_STRING_NOT_ALLOWED",
                message=(
                    "Raw PostgreSQL connection strings are not accepted; use split "
                    "non-secret fields and CARBONOPS_PARSER_POSTGRES_PASSWORD."
                ),
                field_name="CARBONOPS_PARSER_POSTGRES_CONNECTION_STRING",
            ),
        )

    return ProductionConfigValidationResult(issues=tuple(issues))


def _validate_port(
    raw_value: str | None,
    issues: list[ProductionConfigValidationIssue],
) -> None:
    if not _has_text(raw_value):
        return
    try:
        port = int(raw_value)
    except ValueError:
        port = 0
    if not 1 <= port <= 65535:
        issues.append(
            ProductionConfigValidationIssue(
                code="PRODUCTION_CONFIG_INVALID_POSTGRES_PORT",
                message="CARBONOPS_PARSER_POSTGRES_PORT must be an integer between 1 and 65535.",
                field_name="CARBONOPS_PARSER_POSTGRES_PORT",
            ),
        )


def _validate_log_level(
    raw_value: str | None,
    issues: list[ProductionConfigValidationIssue],
) -> None:
    if not _has_text(raw_value):
        return
    if raw_value.strip().lower() not in _VALID_LOG_LEVELS:
        issues.append(
            ProductionConfigValidationIssue(
                code="PRODUCTION_CONFIG_INVALID_LOG_LEVEL",
                message=(
                    "CARBONOPS_PARSER_LOG_LEVEL must be one of debug, info, "
                    "warning, error, or critical."
                ),
                field_name="CARBONOPS_PARSER_LOG_LEVEL",
            ),
        )


def _has_text(value: str | None) -> bool:
    return isinstance(value, str) and bool(value.strip())

